### ISSUE TITLE
EdmondsBlossomShrinking: Possibly very strange bug.

### ISSUE THREAD

**Issue by mibadim** [jgrapht/jgrapht#217]
Consider the Bomb Graph B<sub>n</sub>, that is a Complete Graph C<sub>n</sub> with n additional antenna nodes, each of these antenna nodes connected to exactly one node of the core. A B<sub>5</sub> looks like this:

![bildschirmfoto - 14 07 2016 - 13 05 09](https://cloud.githubusercontent.com/assets/5670523/16837490/e9836eac-49c3-11e6-83e1-21453893a3f4.png)

Obviously, a B<sub>n</sub> has a maximum (even a perfect) matching of size n. So let's create a B<sub>n</sub> as a JUnit test case:

```Java
package test;

import static org.junit.Assert.*;

import java.util.Set;

import org.jgrapht.alg.EdmondsBlossomShrinking;
import org.jgrapht.alg.interfaces.MatchingAlgorithm;
import org.jgrapht.graph.DefaultEdge;
import org.jgrapht.graph.SimpleGraph;
import org.junit.Test;

public class BombGraphTest {
	static int BOMBGRAPH_SIZE = 130;

	@Test
	public void independendTest() {
		SimpleGraph<Integer, DefaultEdge> graph = new SimpleGraph(
				DefaultEdge.class);

		for (int i = 0; i < BOMBGRAPH_SIZE; i++) {
			graph.addVertex(i); // (*)
		}

		for (int i = 0; i < BOMBGRAPH_SIZE; i++) {
			graph.addVertex(i + BOMBGRAPH_SIZE); // (**)
			graph.addEdge(i, i + BOMBGRAPH_SIZE);
		} 
		
		for (int i = 0; i < BOMBGRAPH_SIZE; i++) {
			for (int j = i + 1; j < BOMBGRAPH_SIZE; j++) {
				graph.addEdge(i, j);
			}
		}

		MatchingAlgorithm<Integer, DefaultEdge> maxAlg = new EdmondsBlossomShrinking<>(
				graph);
		Set<DefaultEdge> maxMatching = maxAlg.getMatching();
		// Passes at BOMBGRAPH_SIZE < 130, but fails at BOMBGRAPH_SIZE == 130:
		assertTrue(maxMatching.size() == BOMBGRAPH_SIZE); 
	}
}
```
The assertion passes for `BOMBGRAPH_SIZE < 130`, but fails at `BOMBGRAPH_SIZE >=130`. _However_, when I move the line `(**)` directly below line `(*)`, the test passes for even larger Graphs. What's wrong here?

**Comment by jsichi** [comment]
I'm guessing that it's a combination of 

http://stackoverflow.com/questions/5277881/why-arent-integers-cached-in-java

plus the fact that EdmondsBlossomShrinking is incorrectly using == instead of equals in multiple places.

**Comment by besza** [comment]
#275

**Comment by jkinable** [comment]
This issue has been resolved in #275 and can be closed?

**Comment by jsichi** [comment]
Yep.
